src/System.IO.Compression/src/System/IO/Compression/InflaterZlib.cs

```diff
@@ -0,0 +1,275 @@
+        {
+            return _zlibStream.AvailIn == 0;
+        }
+
+        public void SetFileFormatReader(IFileFormatReader reader)
+        {
+            // Rather than reading raw data and using a FormatReader to interpret
+            // headers/footers manually, we instead reset the zlib stream to parse
+            // that information for us.
+            ZLibNative.ErrorCode error;
+            try
+            {
+                error = _zlibStream.InflateReset(reader.ZLibWindowSize);
+            }
+            catch (Exception exception) // could not load the ZLib DLL
+            {
+                throw new ZLibException(SR.ZLibErrorDLLLoadError, exception);
+            }
+            switch (error)
+            {
+                case ZLibNative.ErrorCode.Ok:           // Successful initialization
+                    return;
+
+                case ZLibNative.ErrorCode.MemError:     // Not enough memory
+                    throw new ZLibException(SR.ZLibErrorNotEnoughMemory, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+
+                case ZLibNative.ErrorCode.VersionError: //zlib library is incompatible with the version assumed
+                    throw new ZLibException(SR.ZLibErrorVersionMismatch, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+
+                case ZLibNative.ErrorCode.StreamError:  // Parameters are invalid
+                    throw new ZLibException(SR.ZLibErrorIncorrectInitParameters, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+
+                default:
+                    throw new ZLibException(SR.ZLibErrorUnexpected, "inflateInit2_", (int)error, _zlibStream.GetErrorMessage());
+            }
+        }
+
+        public void SetInput(byte[] inputBuffer, int startIndex, int count)
+        {
+            Debug.Assert(((IInflater)this).NeedsInput(), "We have something left in previous input!");
+            Debug.Assert(inputBuffer != null);
+            Debug.Assert(startIndex >= 0 && count >= 0 && count + startIndex <= inputBuffer.Length);
+            Debug.Assert(!_inputBufferHandle.HasValue);
+
+            if (0 == count)
+                return;
+
+            lock (_syncLock)
+            {
+                _inputBufferHandle = GCHandle.Alloc(inputBuffer, GCHandleType.Pinned);
+                _zlibStream.NextIn = _inputBufferHandle.Value.AddrOfPinnedObject() + startIndex;
+                _zlibStream.AvailIn = (uint)count;
+                _finished = false;
+            }
+        }
+
+        [SecuritySafeCritical]
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_isDisposed && disposing)
+            {
+                if (_inputBufferHandle.HasValue)
+                    DeallocateInputBufferHandle();
+
+                _zlibStream.Dispose();
+                _isDisposed = true;
+            }
+        }
+
+        void IDisposable.Dispose()
+        {
+            Dispose(true);
+        }
```

Comment:
There is not a finalizer. I will add one and split the code up like so:
```
if (!_isDisposed)
{
    if (disposing)
    {
        _zlibStream.Dispose();
    }
    if (_inputBufferHandle.IsAllocated)
        DeallocateInputBufferHandle();

    _isDisposed = true;
}
```

DeflateStream disposal is not calling the InflateStream's dispose method; it only calls disposal on the _deflater. I'll add some disposals for the inflater in the same place the deflater is being disposed.